### PR TITLE
Fix segfault in jody hash

### PR DESCRIPTION
**Pull request by yerden**
One application of mine has suddenly crashed after running for weeks without a glitch. The system is Intel Xeon E5-2620 v4.

Here's relevant part of the backtrace:
```
goroutine 15 [running]:
runtime.throw(0x855148, 0x5)
        /usr/local/go/src/runtime/panic.go:1114 +0x72 fp=0xc00004c930 sp=0xc00004c900 pc=0x435c22
runtime.sigpanic()
        /usr/local/go/src/runtime/signal_unix.go:702 +0x3cc fp=0xc00004c960 sp=0xc00004c930 pc=0x44b82c
github.com/segmentio/fasthash/jody.AddString64(0x209c71d594e51d87, 0xc0c3fffffa, 0x3, 0x209c71d594e51d87)
        /root/go/pkg/mod/github.com/segmentio/fasthash@v1.0.1/jody/hash.go:102 +0x74 fp=0xc00004c968 sp=0xc00004c960 pc=0x6f25b4
github.com/segmentio/stats/v4/prometheus.labels.hash(0xc0a3df4400, 0x4, 0x8, 0x0)
        /root/go/pkg/mod/github.com/segmentio/stats/v4@v4.5.3/prometheus/label.go:52 +0x82 fp=0xc00004c9a8 sp=0xc00004c968 pc=0x6f5442
github.com/segmentio/stats/v4/prometheus.(*metricEntry).lookup(0xc0001e6090, 0xc0a3df4400, 0x4, 0x8, 0x791450)
        /root/go/pkg/mod/github.com/segmentio/stats/v4@v4.5.3/prometheus/metric.go:166 +0x4c fp=0xc00004ca20 sp=0xc00004c9a8 pc=0x6f65dc
github.com/segmentio/stats/v4/prometheus.(*metricStore).update(0xddb988, 0x1, 0xc08c3fa1a0, 0x6, 0x791450, 0x7, 0x0, 0x0, 0x3ff0000000000000, 0xbfab90f0eb145aab, ...)
        /root/go/pkg/mod/github.com/segmentio/stats/v4@v4.5.3/prometheus/metric.go:100 +0xe4 fp=0xc00004cae8 sp=0xc00004ca20 pc=0x6f60f4
github.com/segmentio/stats/v4/prometheus.(*Handler).HandleMeasures(0xddb960, 0xbfab90f0eb145aab, 0x1a6b4e897307, 0xddba20, 0xc08d884000, 0x1, 0x20)
        /root/go/pkg/mod/github.com/segmentio/stats/v4@v4.5.3/prometheus/handler.go:76 +0x5ed fp=0xc00004cd38 sp=0xc00004cae8 pc=0x6f44dd
github.com/segmentio/stats/v4.(*Engine).ReportAt(0xc000067dc0, 0xbfab90f0eb145aab, 0x1a6b4e897307, 0xddba20, 0x7e9440, 0xc08920b1a0, 0xc00012c000, 0x4, 0x8)
        /root/go/pkg/mod/github.com/segmentio/stats/v4@v4.5.3/engine.go:207 +0x22e fp=0xc00004ce28 sp=0xc00004cd38 pc=0x6ebd9e
github.com/segmentio/stats/v4.(*Engine).Report(0xc000067dc0, 0x7e9440, 0xc08920b1a0, 0xc00012c000, 0x4, 0xa)
        /root/go/pkg/mod/github.com/segmentio/stats/v4@v4.5.3/engine.go:183 +0x82 fp=0xc00004ce80 sp=0xc00004ce28 pc=0x6ebb52
```

I haven't been able to reproduce it again since then. As you can see the panic issued by this piece of code here https://github.com/segmentio/fasthash/blob/f3727d1efd1ebf7ce313d35dea4ad7096d92e72d/jody/hash.go#L102 which has comment about potential segfault. :)

I've fixed hash computation on tail data. Unfortunately this leads to some performance degradation but I haven't found a way to avoid it.

```
name                     old time/op    new time/op    delta
Hash64/jody/algorithm-4    12.7ns ± 1%    13.3ns ± 1%  +4.99%  (p=0.000 n=19+20)

name                     old speed      new speed      delta
Hash64/jody/algorithm-4  2.83GB/s ± 2%  2.70GB/s ± 1%  -4.65%  (p=0.000 n=19+20)
```

### CODE REVIEW

File: jody/hash.go [hunk 1]
```diff
@@ -97,9 +97,22 @@ func AddString64(h uint64, s string) uint64 {
 	}
 
 	if n := (r.Len & 7); n != 0 {
-		m := mask64[n]
-		c := constant & m
-		v := *(*uint64)(p) & m // risk of segfault here?
+		c := constant & mask64[n]
+		var v uint64
```
Comment (achille-roussel) [inline jody/hash.go]: To keeping the local style consistent:
```suggestion
		v := uint64(0)
```

**Review by achille-roussel — APPROVED** [review]
The longer the strings the more the time on this code path gets amortized. 5% degradation on hashing small strings for a program that is then correct seems like a good trade off, thanks for the fix!

**Comment by jnjackins** [comment]
Hi - I'm running into a similar issue with the new code. In my case I can reproduce very reliably (both in the earlier version, and this version).

```
 fatal error: checkptr: pointer arithmetic result points to invalid allocation
 
 goroutine 161 [running]:
 runtime.throw(0x11204c7, 0x40)
 	/usr/local/go/src/runtime/panic.go:1116 +0x72 fp=0xc000071340 sp=0xc000071310 pc=0x46a4d2
 runtime.checkptrArithmetic(0xc00020f510, 0xc0000713f0, 0x1, 0x1)
 	/usr/local/go/src/runtime/checkptr.go:43 +0xbe fp=0xc000071370 sp=0xc000071340 pc=0x4397be
 github.com/segmentio/fasthash/jody.AddString64(0x1c26a908921dd5e0, 0xc00020f50c, 0x4, 0x1c26a908921dd5e0)
```

**Comment by yerden** [comment]
@jnjackins could you show the input which caused the panic and backtrace up to jody hash code?

**Pull request merged**